### ISSUE TITLE
Calender view not working with recent software for TYPO3 12

### ISSUE THREAD

**Issue by stweil** [slub/dfg-viewer#327]
Working link with official DFG Viewer (which uses older code) shows all years:
https://dfg-viewer.de/show?tx_dlf%5Bid%5D=https%3A%2F%2Fdigital.ladenburg.world%2Fdata/Ladenburger_Wochenblatt/mets.xml

Test with local installation (based on recent code):
http://localhost/viewer?tx_dlf%5Bid%5D=https%3A%2F%2Fdigital.ladenburg.world%2Fdata/Ladenburger_Wochenblatt/mets.xml
redirects to
http://localhost/viewer?tx_dlf_calendar%5Baction%5D=years&tx_dlf_calendar%5Bcontroller%5D=Calendar&tx_dlf_calendar%5Bdouble%5D=0&tx_dlf_calendar%5Bid%5D=https%3A%2F%2Fdigital.ladenburg.world%2Fdata%2FLadenburger_Wochenblatt%2Fmets.xml&tx_dlf_calendar%5Bpage%5D=1&cHash=2cfa758bd98ef9165daf956e563678f8
which only shows an empty viewer page.

The software was installed on Debian bullseye using the commands from https://github.com/UB-Mannheim/kitodo-presentation/wiki/Testumgebung-mit-Podman#installation-dfg-viewer-with-typo3-12.

**Comment by stweil** [comment]
Not only the anchor mets fails with recent code, but also the mets for a year.

Working link with official DFG Viewer:
https://dfg-viewer.de/show?tx_dlf[id]=https%3A%2F%2Fdigital.ladenburg.world%2Fdata%2FLadenburger_Wochenblatt%2Fmets_1900.xml

Test with local installation:
http://localhost:10080/viewer?tx_dlf[id]=https%3A%2F%2Fdigital.ladenburg.world%2Fdata%2FLadenburger_Wochenblatt%2Fmets_1900.xml
redirects to
http://localhost:10080/viewer?tx_dlf_calendar%5Baction%5D=calendar&tx_dlf_calendar%5Bcontroller%5D=Calendar&tx_dlf_calendar%5Bdouble%5D=0&tx_dlf_calendar%5Bid%5D=https%3A%2F%2Fdigital.ladenburg.world%2Fdata%2FLadenburger_Wochenblatt%2Fmets_1900.xml&tx_dlf_calendar%5Bpage%5D=1&cHash=11eb274d2e6938c0ac3ef410732f309c
which also does not work.

**Comment by stweil** [comment]
A mets file for an issue works with the official DFG Viewer:
https://dfg-viewer.de/show?tx_dlf[id]=https%3A%2F%2Fdigital.ladenburg.world%2Fdata%2FLadenburger_Wochenblatt%2Fmets_19000602.xml

It also works with the local installation:
http://localhost/viewer?tx_dlf[id]=https%3A%2F%2Fdigital.ladenburg.world%2Fdata%2FLadenburger_Wochenblatt%2Fmets_19000602.xml

**Comment by stweil** [comment]
Anchor, year and issue views work with TYPO3 11 if the patches for TYPO3 12 are not applied.

**Comment by beatrycze-volk** [comment]
Is it possible that reason for it is that DFG Viewer is not yet TYPO3 12 compatible?

**Comment by stweil** [comment]
Yes, that's why I used my fork with the pull request #322.

**Comment by stweil** [comment]
Unrelated: I think the square brackets in `tx_dlf[id]` should be encoded. This seems to be a minor bug (probably in Kitodo.Presentation) which exists for a long time now and is still in master.

**Comment by markusweigelt** [comment]
> Yes, that's why I used my fork with the pull request #322 for the test with TYPO3 12. The successful test with TYPO3 11 was based on the official repositories.

I find this a bit confusing, as you're referring to the #322  in this comment rather than in the description of this issue. The description makes it sound like TYPO3 12 was installed with the state from the DFG-Viewer master.

In the #322, I didn't take the calendar view into account, but I could fix that soon. It would be helpful if we could collect the test scenarios so that we can go through all of them at once.

> Unrelated: I think the square brackets in tx_dlf[id] should be encoded. This seems to be a minor bug (probably in Kitodo.Presentation) which exists for a long time now and is still in master.

Don't you think it would be better to create an issue in Kitodo.Presentation if you already suspect this? It will probably get overlooked here.

**Comment by stweil** [comment]
> I find this a bit confusing, as you're referring to the #322 in this comment rather than in the description of this issue. The description makes it sound like TYPO3 12 was installed with the state from the DFG-Viewer master.

Yes, you are right. My only excuse is that I also have to sort things for me and find the reasons for failures in my tests. When I created the issue I did not know that it was related to TYPO3 12 and your pull request.

**Comment by stweil** [comment]
I have now tested the latest code from the main/master branches with TYPO3 12. The problem still exists. Should it be moved to Kitodo.Presentation? People with write access to both repositories can do this without creating a new issue.

**Comment by stweil** [comment]
A request for viewing the anchor of a newspaper https://digital.ladenburg.world/viewer?tx_dlf%5Bid%5D=https%3A%2F%2Fdigital.ladenburg.de%2Fsamples%2FDeutReunP_856399094_19380115_anchor.xml is redirected to the URL https://digital.ladenburg.world/viewer?tx_dlf_calendar%5Baction%5D=years&tx_dlf_calendar%5Bcontroller%5D=Calendar&tx_dlf_calendar%5Bdouble%5D=0&tx_dlf_calendar%5Bid%5D=https%3A%2F%2Fdigital.ladenburg.de%2Fsamples%2FDeutReunP_856399094_19380115_anchor.xml&tx_dlf_calendar%5Bpage%5D=1&cHash=0ee01a1699221c84e0abd6281384f55e which does not show any content.

**Comment by stweil** [comment]
A request for viewing a year of a newspaper https://digital.ladenburg.world/viewer?tx_dlf%5Bid%5D=https%3A%2F%2Fdigital.ladenburg.de%2Fsamples%2FDeutReunP_856399094_19380115_year.xml is redirected to the URL https://digital.ladenburg.world/viewer?tx_dlf_calendar[action]=years&tx_dlf_calendar[controller]=Calendar&tx_dlf_calendar[id]=https%3A%2F%2Fdigital.ladenburg.de%2Fsamples%2FDeutReunP_856399094_19380115_anchor.xml which does not show any content.

**Comment by stweil** [comment]
CC @sebastian-meyer. This is one of the most annoying bugs preventing the use of the latest Kitodo / DFG-Viewer code with TYPO3 v12.

**Comment by stweil** [comment]
Is it desired to use `tx_dlf_calendar` parameters in the URL while normally `tx_dlf` parameters are used with Kitodo, or is this a bug?

**Comment by sebastian-meyer** [comment]
I am not sure, but I think this is a bug. Generally it is intended that ALL controllers use the same parameter namespace (`tx_dlf`). I don't know why the calendar controller seems to use its own namespace which leads to this issue...

**Comment by markusweigelt** [comment]
We have not defined such a parameter. It seems to be due to a changed behavior of redirects in TYPO3 12, which are triggered by the controller of the Calendar plugin. 

https://github.com/kitodo/kitodo-presentation/blob/4e4a82fa07c4218e071de62b1b8c533279b42dac/Classes/Controller/CalendarController.php#L88

We should take a closer look at why the redirect is no longer working.

**Comment by stweil** [comment]
Maybe the code should use a redirect similar to the ones in Classes/Controller/PageViewController.php?

**Comment by stweil** [comment]
https://github.com/kitodo/kitodo-presentation/pull/1504 seems to fix this issue.

**Comment by markusweigelt** [comment]
> ~Maybe the code should use a redirect similar to the ones in Classes/Controller/PageViewController.php?~
> 
> See https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/11.0/Deprecation-92815-ActionControllerForward.html.

That doesn’t really help here. I’ve already checked the changelogs. No forward function is called, but rather a redirect function. This is more about handling the ResponseInterface, which is returned in TYPO3 12.

Did you check your solution with the complete calendar cause the `requestData` provided to the ForwardResponse?

**Comment by stweil** [comment]
Calling the redirect function was the wrong fix for the deprecated forward function which was used in the old code.

I use the fix for https://digital.ladenburg.world/viewer?tx_dlf[id]=https%3A%2F%2Fdigital.ladenburg.de%2Fdata%2FLadenburger_Wochenblatt%2Fmets.xml.

**Comment by stweil** [comment]
Fixed in latest Kitodo, so closing this issue.
